From: Jason Rhinelander <jason@imaginary.ca>
Date: Sun, 5 May 2019 01:27:17 -0300
Subject: Make gyuanx build and install shared libs

Patch the gyuanx cmake files to properly build and install shared libs.
---
 CMakeLists.txt              | 17 +++++++++++++++--
 contrib/epee/CMakeLists.txt |  6 ++++++
 external/CMakeLists.txt     |  8 +++++++-
 src/CMakeLists.txt          |  6 +++++-
 4 files changed, 33 insertions(+), 4 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index c3aabe5..79dea27 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -44,7 +44,7 @@ if(CCACHE_PROGRAM)
   endforeach()
 endif()
 
-cmake_minimum_required(VERSION 3.10)
+cmake_minimum_required(VERSION 3.13)
 message(STATUS "CMake version ${CMAKE_VERSION}")
 
 # Has to be set before `project()`, and ignored on non-macos:
@@ -335,7 +335,9 @@ endif()
 
 include(cmake/check_for_std_filesystem.cmake)
 
-include_directories(external/rapidjson/include src external)
+include(Version)
+
+include_directories(external/rapidjson/include src contrib/epee/include external)
 
 if(APPLE)
   include_directories(SYSTEM /usr/include/malloc)
@@ -945,6 +947,17 @@ if(BUILD_DOCUMENTATION)
   endif()
 endif()
 
+set(gyuanxlibs epee blockchain_db checkpoints common cncrypto cryptonote_basic
+        cryptonote_core cryptonote_protocol device device_trezor lmdb_lib mnemonics
+        multisig net p2p ringct_basic ringct
+        rpc_server_base rpc rpc_http_client rpc_commands daemon_messages daemon_rpc_server
+        serialization wallet easylogging lmdb randomx blocks version)
+install(TARGETS ${gyuanxlibs}
+    ARCHIVE DESTINATION lib${GYUANX_INSTALL_LIBDIR_SUFFIX}
+    LIBRARY DESTINATION lib${GYUANX_INSTALL_LIBDIR_SUFFIX})
+set_target_properties(${gyuanxlibs} PROPERTIES
+    VERSION "${PROJECT_VERSION_MAJOR}.${PROJECT_VERSION_MINOR}.${PROJECT_VERSION_PATCH}")
+
 find_package(PythonInterp)
 find_program(iwyu_tool_path NAMES iwyu_tool.py iwyu_tool)
 if (iwyu_tool_path AND PYTHONINTERP_FOUND)
diff --git a/contrib/epee/CMakeLists.txt b/contrib/epee/CMakeLists.txt
index 2af962a..c7a707a 100644
--- a/contrib/epee/CMakeLists.txt
+++ b/contrib/epee/CMakeLists.txt
@@ -28,3 +28,9 @@
 
 add_subdirectory(src)
 target_include_directories(epee PUBLIC include)
+
+install(
+    DIRECTORY include/
+    DESTINATION include${GYUANX_INSTALL_INCLUDEDIR_SUFFIX}
+    COMPONENT development
+    FILES_MATCHING PATTERN "*.h" PATTERN "*.inl")
diff --git a/external/CMakeLists.txt b/external/CMakeLists.txt
index 7a90f57..366f42d 100644
--- a/external/CMakeLists.txt
+++ b/external/CMakeLists.txt
@@ -73,7 +73,7 @@ endif()
 
 
 if(NOT GYUANXMQ_FOUND)
-  message(STATUS "Using in-tree lokimq")
+  message(FATAL_ERROR "Using in-tree lokimq")
   if (DEPENDS)
     # The depends system builds a static libzmq so set up a target gyuanx-mq can use to avoid building
     add_library(libzmq STATIC IMPORTED GLOBAL)
@@ -103,6 +103,12 @@ add_subdirectory(db_drivers)
 add_subdirectory(easylogging++ easyloggingpp)
 add_subdirectory(randomx EXCLUDE_FROM_ALL)
 
+install(
+  DIRECTORY boost
+  DESTINATION include${GYUANX_INSTALL_INCLUDEDIR_SUFFIX}
+  COMPONENT   development
+  FILES_MATCHING PATTERN "*.hpp")
+
 # uSockets doesn't really have a proper build system (just a very simple Makefile) so build it
 # ourselves.
 if (NOT CMAKE_VERSION VERSION_LESS 3.12)
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 05c04a7..93ee994 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -58,7 +58,6 @@ function (gyuanx_add_executable target binary)
   set_property(GLOBAL PROPERTY gyuanx_executable_targets "${exec_tgt}")
 endfunction ()
 
-include(Version)
 add_library(version ${CMAKE_BINARY_DIR}/version.cpp)
 add_dependencies(version genversion)
 
@@ -104,3 +103,8 @@ endif()
 
 add_subdirectory(device)
 add_subdirectory(device_trezor)
+
+install(
+    DIRECTORY   "${CMAKE_CURRENT_SOURCE_DIR}/"
+    DESTINATION "include${GYUANX_INSTALL_INCLUDEDIR_SUFFIX}"
+    FILES_MATCHING PATTERN "*.h" PATTERN "*.hpp")
